probe-rs/src/architecture/arm/ap_v2/root_memory_interface.rs

```diff
@@ -0,0 +1,128 @@
+        let base = base_ptr0
+            .valid()
+            .then(|| u64::from(base_ptr1.ptr()) | u64::from(base_ptr0.ptr() << 12))
+            .inspect(|base| tracing::info!("DPv3 BASE_PTR: 0x{base:x}"))
+            .ok_or_else(|| ArmError::Other("DP has no valid base address defined.".into()))?;
+
+        Ok(Self { iface, dp, base })
+    }
+}
+
+impl MemoryInterface<ArmError> for RootMemoryInterface<'_> {
+    fn supports_native_64bit_access(&mut self) -> bool {
+        false
+    }
+
+    fn read_64(&mut self, _address: u64, _data: &mut [u64]) -> Result<(), ArmError> {
+        unimplemented!("The DPv3 only supports 32bit accesses")
+    }
+
+    fn read_32(&mut self, address: u64, data: &mut [u32]) -> Result<(), ArmError> {
+        //tracing::debug!("Reading {} words at {:x} on Root Access Port", data.len(), address);
+        // read content
+        for (i, d) in data.iter_mut().enumerate() {
+            let addr = address + (i as u64) * 4;
+            let base = (self.base + addr) & 0xFFFF_FFFF_FFFF_FFF0;
+            let fqa = FullyQualifiedApAddress::v2_with_dp(self.dp, ApV2Address::new_with_tip(base));
+
+            *d = self.iface.read_raw_ap_register(&fqa, (addr & 0xF) as u8)?;
+        }
+        Ok(())
+    }
+
+    fn read_16(&mut self, _address: u64, _data: &mut [u16]) -> Result<(), ArmError> {
+        unimplemented!("The DPv3 only supports 32bit accesses")
+    }
+
+    fn read_8(&mut self, _address: u64, _data: &mut [u8]) -> Result<(), ArmError> {
+        unimplemented!("The DPv3 only supports 32bit accesses")
+    }
+
+    fn write_64(&mut self, _address: u64, _data: &[u64]) -> Result<(), ArmError> {
+        unimplemented!("The DPv3 only supports 32bit accesses")
+    }
+
+    fn write_32(&mut self, address: u64, data: &[u32]) -> Result<(), ArmError> {
+        // read content
+        for (i, d) in data.iter().enumerate() {
+            let addr = address + (i as u64) * 4;
+            let base = (self.base + addr) & 0xFFFF_FFFF_FFFF_FFF0;
+            let fqa = FullyQualifiedApAddress::v2_with_dp(self.dp, ApV2Address::new_with_tip(base));
+
+            self.iface
+                .write_raw_ap_register(&fqa, (addr & 0xF) as u8, *d)?;
+        }
+        Ok(())
+    }
+
+    fn write_16(&mut self, _address: u64, _data: &[u16]) -> Result<(), ArmError> {
+        unimplemented!("The DPv3 only supports 32bit accesses")
+    }
+
+    fn write_8(&mut self, _address: u64, _data: &[u8]) -> Result<(), ArmError> {
+        unimplemented!("The DPv3 only supports 32bit accesses")
+    }
+
+    fn supports_8bit_transfers(&self) -> Result<bool, ArmError> {
+        Ok(false)
+    }
+
+    fn flush(&mut self) -> Result<(), ArmError> {
+        // transfers are not buffered.
+        Ok(())
+    }
+}
+impl ArmMemoryInterface for RootMemoryInterface<'_> {
+    fn fully_qualified_address(&self) -> FullyQualifiedApAddress {
+        FullyQualifiedApAddress::v2_with_dp(self.dp, ApV2Address::new())
```

Comment:
It seems weird that this is implemented here. From what I understand, this is used to access APs, but is not realted to an actual AP. So it seems wrong that this returns an AP address.

Reply:
I think the differentiation is that it returns an `Apv2Address` that corresponds with "nothing" (i.e. an empty vector). I agree it's a bit odd.

Reply:
Is `ApV2Address` with an empty vector referring to nothing, or the "root AP"? See https://github.com/probe-rs/probe-rs/pull/3005#discussion_r1908133149.

Regardless, I would prefer a clearer function name, e.g. something like `ApV2Address::root()` (if it's actually referring to the "root AP").

Reply:
Yes, what’s returned by `ApV2Address::new()` is the root AP (the DP’s memory space).

Reply:
So this is kind of a "virtual AP", this is not an AP which really exists? I.e. its not an actual `MemoryAp` as described in the spec.

Reply:
I wouldn't call the root memory interface an AP (its really the DP memory space), but it does have a `FullyQualifiedApAddress` associated with it because it's used in the chain of ports to access a final address.